{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Script to analyze the multiple myeloma results. This includes plotting the representative tree and gene expression analysis.",
   "id": "88f416f867ed1873"
  },
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-05-12T17:25:21.308333Z",
     "start_time": "2025-05-12T17:25:21.292204Z"
    }
   },
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.cm as cm\n",
    "from IPython.core.pylabtools import figsize\n",
    "from networkx.algorithms.bipartite.cluster import clustering\n",
    "from torchgen.gen_functionalization_type import return_str\n",
    "\n",
    "from src_python.cell_tree import CellTree"
   ],
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-12T17:25:21.337861Z",
     "start_time": "2025-05-12T17:25:21.325903Z"
    }
   },
   "cell_type": "code",
   "source": [
    "study_num = \"bc03\" #\"mm34\" #\"bc03\"\n",
    "primary_cells = []\n",
    "\n",
    "if study_num == \"bc03\":\n",
    "    primary_cells = [\"SRR5023442\", \"SRR5023443\", \"SRR5023444\", \"SRR5023445\", \"SRR2973275\", \"SRR2973351\", \"SRR2973352\", \"SRR2973353\", \"SRR2973354\", \"SRR2973355\", \"SRR2973356\", \"SRR2973357\", \"SRR2973358\", \"SRR2973359\", \"SRR2973360\", \"SRR2973361\", \"SRR2973362\", \"SRR2973363\", \"SRR2973364\", \"SRR2973365\", \"SRR2973366\", \"SRR2973367\", \"SRR2973368\", \"SRR2973369\", \"SRR2973370\", \"SRR2973371\", \"SRR2973372\", \"SRR2973373\", \"SRR2973374\", \"SRR2973375\", \"SRR2973376\", \"SRR2973377\", \"SRR2973378\", \"SRR2973379\", \"SRR2973380\", \"SRR2973381\", \"SRR2973382\", \"SRR2973383\"]\n",
    "\n",
    "if study_num == \"bc07\":\n",
    "    primary_cells =  [\"SRR2973484\", \"SRR5023558\", \"SRR5023559\", \"SRR5023560\", \"SRR2973277\", \"SRR2973437\", \"SRR2973438\", \"SRR2973439\", \"SRR2973440\", \"SRR2973441\", \"SRR2973442\", \"SRR2973443\", \"SRR2973444\", \"SRR2973445\", \"SRR2973446\", \"SRR2973447\", \"SRR2973448\", \"SRR2973449\", \"SRR2973450\", \"SRR2973451\", \"SRR2973452\", \"SRR2973453\", \"SRR2973454\", \"SRR2973455\", \"SRR2973456\", \"SRR2973457\", \"SRR2973458\", \"SRR2973459\", \"SRR2973460\", \"SRR2973461\", \"SRR2973462\", \"SRR2973463\", \"SRR2973464\", \"SRR2973465\", \"SRR2973466\", \"SRR2973467\", \"SRR2973468\", \"SRR2973469\", \"SRR2973470\", \"SRR2973471\", \"SRR2973472\", \"SRR2973473\", \"SRR2973474\", \"SRR2973475\", \"SRR2973476\", \"SRR2973477\", \"SRR2973478\", \"SRR2973479\", \"SRR2973480\", \"SRR2973481\", \"SRR2973482\", \"SRR2973483\"]\n",
    "\n",
    "if study_num == \"mm34\":\n",
    "    primary_cells = [\"SRR6710302\", \"SRR6710303\", \"SRR6710304\", \"SRR6710305\", \"SRR6710306\", \"SRR6710307\", \"SRR6710308\", \"SRR6710309\", \"SRR6710310\", \"SRR6710311\", \"SRR6710312\", \"SRR6710313\", \"SRR6710314\", \"SRR6710315\", \"SRR6710316\", \"SRR6710317\", \"SRR6710318\", \"SRR6710319\", \"SRR6710320\", \"SRR6710321\", \"SRR6710322\", \"SRR6710323\", \"SRR6710324\", \"SRR6710325\", \"SRR6710326\", \"SRR6710327\", \"SRR6710328\", \"SRR6710329\", \"SRR6710330\", \"SRR6710331\", \"SRR6710332\", \"SRR6710333\", \"SRR6710334\", \"SRR6710335\", \"SRR6710336\", \"SRR6710337\", \"SRR6710338\", \"SRR6710339\", \"SRR6710340\", \"SRR6710341\", \"SRR6710342\", \"SRR6710343\", \"SRR6710344\", \"SRR6710345\", \"SRR6710346\", \"SRR6710347\", \"SRR6710348\", \"SRR6710349\", \"SRR6710350\", \"SRR6710351\", \"SRR6710352\", \"SRR6710353\", \"SRR6710354\", \"SRR6710355\", \"SRR6710356\", \"SRR6710357\", \"SRR6710358\", \"SRR6710359\", \"SRR6710360\", \"SRR6710361\", \"SRR6710362\", \"SRR6710363\", \"SRR6710364\", \"SRR6710365\", \"SRR6710366\"]\n",
    "\n",
    "if study_num == \"mm16\":\n",
    "    primary_cells = [\"SRR6710256\", \"SRR6710257\", \"SRR6710258\", \"SRR6710259\", \"SRR6710260\", \"SRR6710261\", \"SRR6710262\", \"SRR6710263\", \"SRR6710264\", \"SRR6710265\", \"SRR6710266\", \"SRR6710267\", \"SRR6710268\", \"SRR6710269\", \"SRR6710270\", \"SRR6710271\", \"SRR6710272\", \"SRR6710273\", \"SRR6710274\", \"SRR6710275\", \"SRR6710276\", \"SRR6710277\", \"SRR6710278\"]"
   ],
   "id": "9e0d8e7fd0673e6b",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-12T17:25:23.685077Z",
     "start_time": "2025-05-12T17:25:21.356870Z"
    }
   },
   "cell_type": "code",
   "source": [
    "n_bootstrap = 1000\n",
    "np.random.seed(0)\n",
    "n_rounds = 3\n",
    "model = \"sclineager\"   # \"sciterna\"\n",
    "\n",
    "path = rf\"../data/results/{study_num}/{model}\"\n",
    "os.makedirs(os.path.join(path, f\"{model}_ct_trees\"), exist_ok=True)\n",
    "\n",
    "ref = np.array(pd.read_csv(os.path.join(f\"../data/input_data/{study_num}\", \"ref.csv\")))\n",
    "cell_names = list(ref[:,0])\n",
    "cell_indices = [cell_names.index(item) for item in primary_cells if item in cell_names]\n",
    "ref = ref[:, 1:]\n",
    "\n",
    "alt = np.array(pd.read_csv(os.path.join(f\"../data/input_data/{study_num}\", \"alt.csv\")))[:, 1:]\n",
    "\n",
    "def plot_trees(path, round, plot_mutations=False, min_num=0):\n",
    "    for i in range(min_num, n_rounds):\n",
    "        path_parent = os.path.join(path, f\"{model}_parent_vec\", f\"{model}_parent_vec_{i}{round}.txt\")\n",
    "        path_mut_loc = os.path.join(path, f\"{model}_mutation_location\", f\"{model}_mutation_location_{i}{round}.txt\")\n",
    "        path_selected = os.path.join(path, f\"{model}_selected_loci\", f\"{model}_selected_loci_{i}{round}.txt\")\n",
    "\n",
    "        parent_vec = np.loadtxt(path_parent, dtype=int)\n",
    "        if os.path.exists(path_mut_loc) == False:\n",
    "            mut_locs = []\n",
    "            selected_mutations = []\n",
    "        else:\n",
    "            mut_locs = np.loadtxt(path_mut_loc, dtype=int)\n",
    "            selected_mutations = np.loadtxt(path_selected, dtype=int)\n",
    "\n",
    "        n_cells = int(((len(parent_vec)+1)/2))\n",
    "        ct = CellTree(n_cells=n_cells, n_mut=len(selected_mutations))\n",
    "        ct.use_parent_vec(parent_vec)\n",
    "\n",
    "        if plot_mutations:\n",
    "            ct.mut_loc = mut_locs\n",
    "            graph = ct.to_graphviz()\n",
    "            graph.render(os.path.join(path, f\"{model}_ct_trees\", f\"ct_tree_{i}_{round}\"), format='png', cleanup=True)\n",
    "\n",
    "        ct.mut_loc = []\n",
    "        graph_branches = ct.to_graphviz()\n",
    "        for n in range(n_cells):\n",
    "            if study_num == \"mm34\":\n",
    "                if n in cell_indices:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"red\")\n",
    "                else:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"blue\") # metastasis\n",
    "            elif study_num == \"mm16\":\n",
    "                if n < 23:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"red\")\n",
    "                else:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"blue\") # after treatment\n",
    "            elif study_num == \"bc03\":\n",
    "                if n in cell_indices:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"red\")\n",
    "                else:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"blue\") # lymph node metastasis\n",
    "            elif study_num == \"bc07\":\n",
    "                if n in cell_indices:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"red\")\n",
    "                else:\n",
    "                    graph_branches.node(str(n), shape='circle', style='filled', color=\"blue\") # lymph node metastasis\n",
    "\n",
    "\n",
    "        graph_branches.render(os.path.join(path, f\"{model}_ct_trees\", f\"ct_tree_{i}_{round}_branches\"), format='png', cleanup=True)\n",
    "\n",
    "if model == \"sciterna\":\n",
    "    plot_trees(path, f\"0r\", plot_mutations=True)\n",
    "    path_bootstrap = rf\"../data/results/{study_num}/{model}_bootstrap\"\n",
    "    os.makedirs(os.path.join(path_bootstrap, f\"{model}_ct_trees\"), exist_ok=True)\n",
    "    for i in range(n_bootstrap):\n",
    "        plot_trees(path_bootstrap, f\"{i}r\", plot_mutations=False)\n",
    "else:\n",
    "    plot_trees(path, \"\", plot_mutations=True, min_num = 2)"
   ],
   "id": "6db3b6020b68eb6c",
   "outputs": [],
   "execution_count": 27
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-12T17:25:23.719828Z",
     "start_time": "2025-05-12T17:25:23.696281Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import seaborn as sns\n",
    "\n",
    "vaf = alt/(ref+alt)\n",
    "vafq = pd.isna(vaf)\n",
    "missing = np.sum(vafq, axis=1)\n",
    "vaf[vafq] = 0.0\n",
    "vaf = np.array(vaf, dtype=float)\n",
    "\n",
    "plt.figure(figsize=(5,10))\n",
    "sns.heatmap(vaf)\n",
    "plt.title(\"Variant Allele Frequency (VAF) Heatmap\")\n",
    "plt.xlabel(\"Samples\")\n",
    "plt.ylabel(\"Mutations\")\n",
    "plt.show()"
   ],
   "id": "4412ad89acb0d65e",
   "outputs": [
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mZeroDivisionError\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[1;32mIn[28], line 3\u001B[0m\n\u001B[0;32m      1\u001B[0m \u001B[38;5;28;01mimport\u001B[39;00m \u001B[38;5;21;01mseaborn\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m \u001B[38;5;21;01msns\u001B[39;00m\n\u001B[1;32m----> 3\u001B[0m vaf \u001B[38;5;241m=\u001B[39m \u001B[43malt\u001B[49m\u001B[38;5;241;43m/\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43mref\u001B[49m\u001B[38;5;241;43m+\u001B[39;49m\u001B[43malt\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m      4\u001B[0m vafq \u001B[38;5;241m=\u001B[39m pd\u001B[38;5;241m.\u001B[39misna(vaf)\n\u001B[0;32m      5\u001B[0m missing \u001B[38;5;241m=\u001B[39m np\u001B[38;5;241m.\u001B[39msum(vafq, axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m1\u001B[39m)\n",
      "\u001B[1;31mZeroDivisionError\u001B[0m: division by zero"
     ]
    }
   ],
   "execution_count": 28
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "missing",
   "id": "9b6beba909a89012",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "np.argwhere(posterior_node_after_branching == 5)",
   "id": "5bc8dcc618c53a1a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "selected = np.loadtxt(os.path.join(path_bootstrap, \"selected.txt\"), delimiter=',', dtype=int)\n",
    "ref = pd.read_csv(os.path.join(f\"../data/input_data/{study_num}\", \"ref.csv\"))\n",
    "\n",
    "all_individual_overdispersions_h = np.full((n_bootstrap, ref.shape[1]), np.nan)\n",
    "all_individual_dropouts = np.full((n_bootstrap, ref.shape[1]), np.nan)\n",
    "\n",
    "for i in range(n_bootstrap):\n",
    "    for j in range(n_rounds):\n",
    "        selected_mutations = np.loadtxt(os.path.join(path_bootstrap, \"sciterna_selected_loci\", f\"sciterna_selected_loci_{j}r{i}.txt\"), dtype=int)\n",
    "        individual_dropouts = np.loadtxt(os.path.join(path_bootstrap, \"sciterna_individual_dropout_probs\", f\"sciterna_individual_dropout_probs_{j}r{i}.txt\"))\n",
    "        individual_overdispersions_h = np.loadtxt(os.path.join(path_bootstrap, \"sciterna_individual_overdispersions_h\", f\"sciterna_individual_overdispersions_h_{j}r{i}.txt\"))\n",
    "\n",
    "        unique_mutations = np.unique(selected_mutations)\n",
    "        for mut in unique_mutations:\n",
    "            indices = np.where(selected_mutations == mut)[0]\n",
    "            mean_dropout = np.mean(individual_dropouts[indices])\n",
    "            mean_overdispersion = np.mean(individual_overdispersions_h[indices])\n",
    "\n",
    "            all_individual_dropouts[i, mut] = mean_dropout\n",
    "            all_individual_overdispersions_h[i, mut] = mean_overdispersion"
   ],
   "id": "bcd99c88e6cfaff6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "sufficient_data_columns_od = ~(np.any(all_individual_overdispersions_h == 6, axis=0))\n",
    "sufficient_data_columns_dropout = ~(np.any(all_individual_dropouts == 0.2, axis=0))\n",
    "\n",
    "mean_overdispersions = np.nanmean(all_individual_overdispersions_h[:, sufficient_data_columns_od], axis=0)\n",
    "mean_dropouts = np.nanmean(all_individual_dropouts[:, sufficient_data_columns_dropout], axis=0)"
   ],
   "id": "c2494c6ca183eb90",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "plt.hist(mean_overdispersions, bins=100)\n",
    "plt.title(\"Overdispersion Heterozygous distribution\")\n",
    "plt.show()\n",
    "\n",
    "plt.hist(mean_dropouts, bins=100)\n",
    "plt.title(\"Dropout heterozygous distribution\")\n",
    "plt.show()"
   ],
   "id": "4b145c788de1f279",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "path_sclineager = f\"../data/results/{study_num}/sclineager\"\n",
    "path_dendro = f\"../data/results/{study_num}/dendro\"\n",
    "\n",
    "clones = np.arange(2,6)\n",
    "\n",
    "def plot_model_trees(path, model):\n",
    "\n",
    "    os.makedirs(os.path.join(path, f\"{model}_ct_trees\"), exist_ok=True)\n",
    "\n",
    "    for c in clones:\n",
    "        path_parent = os.path.join(path, f\"{model}_parent_vec\", f\"{model}_parent_vec_{c}.txt\")\n",
    "\n",
    "        parent_vec = np.loadtxt(path_parent, dtype=int)\n",
    "\n",
    "        n_cells = int(((len(parent_vec)+1)/2))\n",
    "        ct = CellTree(n_cells=n_cells)\n",
    "        ct.use_parent_vec(parent_vec)\n",
    "\n",
    "        ct.mut_loc = []\n",
    "        graph = ct.to_graphviz()\n",
    "        for n in range(n_cells):\n",
    "            if n < 65:\n",
    "                graph.node(str(n), shape='circle', style='filled', color=\"red\")\n",
    "            elif n >=65:\n",
    "                graph.node(str(n), shape='circle', style='filled', color=\"blue\")\n",
    "        graph.render(os.path.join(path, f\"{model}_ct_trees\", f\"ct_tree_{c}_branches\"), format='png', cleanup=True)\n",
    "\n",
    "plot_model_trees(path_sclineager, \"sclineager\")\n",
    "plot_model_trees(path_dendro, \"dendro\")"
   ],
   "id": "37c1c3ce16c6a343",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def create_mutation_matrix(parent_vector, mutation_indices, ct):\n",
    "    n_nodes = len(parent_vector)\n",
    "    n_leaves = int((n_nodes+1)/2)\n",
    "    n_mutations = len(mutation_indices)\n",
    "\n",
    "    mutation_matrix = np.zeros((n_nodes, n_mutations), dtype=int)\n",
    "\n",
    "    for mutation_idx, cell_idx in enumerate(mutation_indices):\n",
    "        children = [c for c in ct.dfs(cell_idx)]\n",
    "        for cell in children:  # Traverse all cells below the mutation cell\n",
    "            mutation_matrix[cell, mutation_idx] = 1  # Mark cells with the mutation\n",
    "\n",
    "    return mutation_matrix[:n_leaves].T\n",
    "\n",
    "def are_all_arrays_identical(arrays_list):\n",
    "    first_array = arrays_list[0]\n",
    "    return all(np.array_equal(first_array, array) for array in arrays_list[1:])"
   ],
   "id": "b6c7c8685fd6b65b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# with open(\"../data/input_data/gencode.v21.annotation.gtf\", 'r') as file:\n",
    "#     for line in file:\n",
    "#         if line.startswith(\"#\"):\n",
    "#                 continue\n",
    "#         fields = line.strip().split('\\t')\n",
    "#         if fields[2] == 'gene':\n",
    "#             attributes = {key_value.split(' ')[0]: key_value.split(' ')[1].strip('\"') for key_value in\n",
    "#                           fields[8].split('; ') if key_value}\n",
    "#             gene_names.append([fields[0], fields[3], fields[4], attributes[\"gene_name\"]])\n",
    "# \n",
    "# df = pd.DataFrame(gene_names, columns=['chromosome', 'start', 'end', 'gene'])\n",
    "# df['start'] = df['start'].astype(int)\n",
    "# df['end'] = df['end'].astype(int)\n",
    "# df.to_csv(\"../data/input_data/mm34/gene_positions.csv\", index=False)\n",
    "\n",
    "def convert_location_to_gene(locs):\n",
    "\n",
    "    loc_to_gene = []\n",
    "\n",
    "    df = pd.read_csv(\"../data/input_data/mm34/gene_positions.csv\", index_col=False)\n",
    "    for location in locs:\n",
    "        chrom, pos = location.split(\":\")[0], location.split(\":\")[1]\n",
    "        pos = int(pos)\n",
    "        matching_rows = df[(df['chromosome'] == chrom) & (df['start'] <= pos) & (df['end'] >= pos)]\n",
    "        matching_genes = matching_rows['gene'].tolist()\n",
    "        loc_to_gene.append(matching_genes)\n",
    "\n",
    "    return loc_to_gene"
   ],
   "id": "ff301ec89835b3b4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "n_bootstrap = 1000\n",
    "np.random.seed(0)\n",
    "use_summary_statistics = False # Uses the summary files saved in the data/ directory. If raw output files were generated, set to False\n",
    "\n",
    "path = rf\"../data/results/{study_num}/sciterna_bootstrap\"\n",
    "\n",
    "reference = pd.read_csv(rf'../data/input_data/{study_num}/ref.csv', index_col=0)\n",
    "\n",
    "if use_summary_statistics == False:\n",
    "    os.makedirs(os.path.join(path, \"sciterna_mutations_before_branching\"), exist_ok=True)\n",
    "    os.makedirs(os.path.join(path, \"sciterna_nodes_after_branching\"), exist_ok=True)\n",
    "    os.makedirs(os.path.join(path, \"sciterna_nodes_before_branching\"), exist_ok=True)\n",
    "    os.makedirs(os.path.join(path, \"sciterna_mutations_branching\"), exist_ok=True)\n",
    "    os.makedirs(os.path.join(path, \"sciterna_mutations_after_branching\"), exist_ok=True)\n",
    "    \n",
    "    all_loci = reference.columns\n",
    "    all_cells = reference.index\n",
    "    mut_indicator = {}\n",
    "    \n",
    "    def mutations_node(parent_vector, mutation_indices, ct, node):\n",
    "        n_cells = len(parent_vector)\n",
    "        n_mutations = len(mutation_indices)\n",
    "    \n",
    "        mutation_matrix = np.zeros((n_cells, n_mutations), dtype=int)\n",
    "    \n",
    "        for mutation_idx, cell_idx in enumerate(mutation_indices):\n",
    "            children = [c for c in ct.dfs_experimental(cell_idx)]\n",
    "            for cell in children:  # Traverse all cells below the mutation cell\n",
    "                mutation_matrix[cell, mutation_idx] = 1  # Mark cells with the mutation\n",
    "    \n",
    "        return mutation_matrix[node]\n",
    "    \n",
    "    for i in range(0, n_bootstrap):\n",
    "        for r in range(n_rounds):\n",
    "            path_parent = os.path.join(path, \"sciterna_parent_vec\", f\"sciterna_parent_vec_{r}r{i}.txt\")\n",
    "\n",
    "            if os.path.exists(os.path.join(path, rf\"sciterna_mutations_branching\\mutations_branching_{r}r{i}.txt\")):\n",
    "                continue\n",
    "\n",
    "            path_mut_loc = os.path.join(path, \"sciterna_mutation_location\", f\"sciterna_mutation_location_{r}r{i}.txt\")\n",
    "            path_selected = os.path.join(path, \"sciterna_selected_loci\", f\"sciterna_selected_loci_{r}r{i}.txt\")\n",
    "            parent_vec = np.loadtxt(path_parent, dtype=int)\n",
    "            mut_locs = np.loadtxt(path_mut_loc, dtype=int)\n",
    "            mutations = np.loadtxt(path_selected)\n",
    "            for loci in all_loci:\n",
    "                mut_indicator[loci] = []\n",
    "\n",
    "            n_cells = int(((len(parent_vec)+1)/2))\n",
    "            ct = CellTree(n_cells=n_cells, n_mut=len(mutations))\n",
    "            ct.use_parent_vec(parent_vec)\n",
    "            ct.mut_loc = mut_locs\n",
    "\n",
    "            # mutation_matrix = create_mutation_matrix(ct.parent_vec, ct.mut_loc, ct)\n",
    "            #\n",
    "            # for mi, mut in enumerate(mutations):\n",
    "            #     mut_indicator[all_loci[int(mut)]].append(mutation_matrix[mi])\n",
    "            #\n",
    "            # for k in mut_indicator.keys():\n",
    "            #     if len(mut_indicator[k]) > 1:\n",
    "            #         if not are_all_arrays_identical(mut_indicator[k]):\n",
    "            #             print(\"Problem\")\n",
    "            #         else:\n",
    "            #             mut_indicator[k] = [mut_indicator[k][0]]\n",
    "            #\n",
    "            # dataframe = {}\n",
    "            #\n",
    "            # for key, value in mut_indicator.items():\n",
    "            #     if len(value) > 0:\n",
    "            #         dataframe[key] = pd.DataFrame([value[0]], columns=all_cells)\n",
    "            #     else:\n",
    "            #         dataframe[key] = pd.DataFrame([np.nan] * len(all_cells)).T\n",
    "            #         dataframe[key].columns = all_cells\n",
    "            #\n",
    "            # df = pd.concat(dataframe.values(), axis=0)\n",
    "            # df.index = all_loci\n",
    "            # df.to_csv(os.path.join(path, \"sciterna_mut_indicator\", f\"sciterna_mut_indicator_{r}r{i}.csv\"), index=True)\n",
    "\n",
    "            # optimize capture rate + purity to determine the majority metastasis branch\n",
    "            branching_node = 0\n",
    "            branching_nodes_max = 0\n",
    "            for node in range(len(parent_vec)):\n",
    "                if ct.isleaf(node):\n",
    "                    continue\n",
    "\n",
    "                primary = 0\n",
    "                metastasis = 0\n",
    "                for sn in ct.dfs_experimental(node):\n",
    "                    if ct.isleaf(sn):\n",
    "                        if sn < 65:\n",
    "                            primary += 1\n",
    "                        elif sn >=65:\n",
    "                            metastasis += 1\n",
    "\n",
    "                capture_rate = metastasis/(n_cells - 65)\n",
    "                purity = metastasis/(metastasis + primary)\n",
    "                branching_nodes_new = capture_rate + purity\n",
    "\n",
    "                if branching_nodes_max < branching_nodes_new:\n",
    "                    branching_node = node\n",
    "                    branching_nodes_max = branching_nodes_new\n",
    "            print(branching_node, branching_nodes_max)\n",
    "            if branching_nodes_max < 1.9:\n",
    "                print(\"Potential problem (Likely wrong root): \", i, r)\n",
    "\n",
    "            mutations_indices = mutations_node(parent_vec, mut_locs, ct, branching_node)\n",
    "            mutations_branching = mutations[mutations_indices==1]\n",
    "            np.savetxt(os.path.join(path, rf\"sciterna_mutations_branching\\mutations_branching_{r}r{i}.txt\"), mutations_branching, fmt='%i')\n",
    "\n",
    "            mutations_after_metastasis = []\n",
    "            metastasis_nodes = [n for n in ct.dfs_experimental(branching_node) if ct.isleaf(n)]\n",
    "            primary_nodes = [n for n in range(n_cells) if n not in metastasis_nodes]\n",
    "\n",
    "            np.savetxt(os.path.join(path, rf\"sciterna_nodes_after_branching\\nodes_after_branching_{r}r{i}.txt\"), metastasis_nodes, fmt='%i')\n",
    "            np.savetxt(os.path.join(path, rf\"sciterna_nodes_before_branching\\nodes_before_branching_{r}r{i}.txt\"), primary_nodes, fmt='%i')\n",
    "\n",
    "            for ni, mut_loc in enumerate(mut_locs):\n",
    "                if mut_loc in metastasis_nodes[1:]:\n",
    "                    mutations_after_metastasis.append(mutations[ni])\n",
    "\n",
    "            mutations_before_metastasis = [m for m in mutations if m not in mutations_after_metastasis]\n",
    "\n",
    "            np.savetxt(os.path.join(path, rf\"sciterna_mutations_before_branching\\mutations_before_branching_{r}r{i}.txt\"), mutations_before_metastasis, fmt='%i')\n",
    "            np.savetxt(os.path.join(path, rf\"sciterna_mutations_after_branching\\mutations_after_branching_{r}r{i}.txt\"), mutations_after_metastasis, fmt='%i')"
   ],
   "id": "b5d74b0b910c62ea",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "if use_summary_statistics == False:\n",
    "    primary = []\n",
    "    metastasis = []\n",
    "    primary_nodes = []\n",
    "    metastasis_nodes = []\n",
    "    branching = []\n",
    "    r = 2\n",
    "    for i in range(n_bootstrap):\n",
    "        path_primary_node = os.path.join(path, rf\"sciterna_nodes_before_branching\\nodes_before_branching_{r}r{i}.txt\")\n",
    "        path_metastasis_node = os.path.join(path, rf\"sciterna_nodes_after_branching\\nodes_after_branching_{r}r{i}.txt\")\n",
    "        path_primary = os.path.join(path, rf\"sciterna_mutations_before_branching\\mutations_before_branching_{r}r{i}.txt\")\n",
    "        path_metastasis = os.path.join(path, rf\"sciterna_mutations_after_branching\\mutations_after_branching_{r}r{i}.txt\")\n",
    "        path_branching = os.path.join(path, rf\"sciterna_mutations_branching\\mutations_branching_{r}r{i}.txt\")\n",
    "\n",
    "        primary_data = np.loadtxt(path_primary, dtype=int)\n",
    "        metastasis_data = np.loadtxt(path_metastasis, dtype=int)\n",
    "        primary_nodes_data = np.loadtxt(path_primary_node, dtype=int)\n",
    "        metastasis_nodes_data = np.loadtxt(path_metastasis_node, dtype=int)\n",
    "        branching_data = np.loadtxt(path_branching, dtype=int)\n",
    "\n",
    "        primary.extend(primary_data)\n",
    "        metastasis.extend(metastasis_data)\n",
    "        primary_nodes.extend(primary_nodes_data)\n",
    "        metastasis_nodes.extend(metastasis_nodes_data)\n",
    "        branching.extend(branching_data)\n",
    "        \n",
    "    unique_mutations_branching, counts_mutations_branching = np.unique(np.array(branching), return_counts=True)\n",
    "    selected = np.loadtxt(os.path.join(path, \"selected.txt\"), delimiter=',', dtype=int)\n",
    "    selected_mutations = np.unique(selected)\n",
    "\n",
    "else:\n",
    "    metastasis_nodes = np.loadtxt(os.path.join(path, \"metastasis_nodes.txt\"))"
   ],
   "id": "6b3279aebfb07b47",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "unique_cells, posterior_node_after_branching = np.unique(np.array(metastasis_nodes), return_counts=True)\n",
    "heatmap_data = posterior_node_after_branching.reshape(1, -1)\n",
    "plt.figure(figsize=(30, 2))\n",
    "plt.imshow(heatmap_data, cmap='viridis', aspect='auto')\n",
    "plt.colorbar(label='Count')\n",
    "plt.xticks(ticks=np.arange(len(unique_cells)), labels=posterior_node_after_branching, rotation=90)\n",
    "plt.yticks([])\n",
    "plt.xlabel('Nodes')\n",
    "plt.title('Counts per Cell of Appearing in Majority Metastasis Branch', fontsize=32)\n",
    "plt.grid(False)\n",
    "\n",
    "plt.show()"
   ],
   "id": "387c79c129d0b9a2",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "top_loci = np.loadtxt(rf\"../data/results/{study_num}/sciterna_bootstrap/selected.txt\", dtype=int)",
   "id": "34f0e9320c26a631",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "selected_loci = reference.columns[top_loci]\n",
    "selected_genes = convert_location_to_gene(selected_loci)"
   ],
   "id": "812aae146d344c79",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "genes = []\n",
    "for s in selected_genes:\n",
    "    genes.extend(s)\n",
    "gens, cnt = np.unique(genes, return_counts=True)\n",
    "\n",
    "print(cnt[np.argsort(cnt)][-30:])\n",
    "print(gens[np.argsort(cnt)][-30:])"
   ],
   "id": "1760a8b905acfcf0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "expression_counts_scaled = pd.read_csv(rf\"../data/input_data/{study_num}/normalized_expression_data.csv\", index_col=0)\n",
    "\n",
    "expression_counts_scaled['sum_counts'] = expression_counts_scaled.sum(axis=1)\n",
    "df_sorted = expression_counts_scaled.sort_values(by='sum_counts', ascending=False)\n",
    "top_10000 = df_sorted.head(1000)\n",
    "top_10000 = top_10000.drop(columns=['sum_counts'])\n",
    "expression_counts_scaled = expression_counts_scaled.drop(columns=['sum_counts'])"
   ],
   "id": "8f40bf980cb4bfe1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "top_5_indices = np.argsort(np.mean(expression_counts_scaled, axis=1))[-20:]\n",
    "top_5_rows = expression_counts_scaled.iloc[top_5_indices]\n",
    "top_5_rows"
   ],
   "id": "7e3d8d047514166",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "categories = {\n",
    "    \"Most expressed\": [\"IGKV2-28\", \"B2M\", \"MALAT1\"],\n",
    "    \"largest_difference_groups\": ['IGJ', 'FCRL2', 'CD74', 'ABCA1', 'TBL1XR1', 'HLA-B'], #'CTSB'\n",
    "    \"selected\": ['CTSB'],\n",
    "    \"most_mutated\": ['DCC', 'PELI1', 'NBPF14', 'VPS41', 'ATP11B'] # 'CTSB'\n",
    "}\n",
    "\n",
    "filtered_df_top_genes = pd.DataFrame()\n",
    "\n",
    "for category, genes in categories.items():\n",
    "    category_genes = expression_counts_scaled[expression_counts_scaled.index.isin(genes)]\n",
    "    category_row_sums = category_genes.sum(axis=1)\n",
    "    top_genes = category_row_sums.nlargest(7).index\n",
    "    filtered_df_top_genes = pd.concat([filtered_df_top_genes, expression_counts_scaled.loc[top_genes]])"
   ],
   "id": "7406a18801f20a1a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "sorted_indices = np.argsort(posterior_node_after_branching)\n",
    "mask_less_than_65 = sorted_indices < 65\n",
    "\n",
    "plt.figure(figsize=(40, 16))\n",
    "\n",
    "sorted_df = filtered_df_top_genes.iloc[:, sorted_indices]\n",
    "col_colors = ['red' if val else 'blue' for val in mask_less_than_65]\n",
    "\n",
    "plt.figure(figsize=(40, 18))\n",
    "ax = sns.heatmap(sorted_df, cmap='coolwarm')\n",
    "\n",
    "for n, tick_label in enumerate(ax.get_xticklabels()):\n",
    "    tick_label.set_color(col_colors[n])\n",
    "        \n",
    "for tick_label in ax.get_yticklabels():\n",
    "    tick_label.set_fontsize(26)   \n",
    "    tick_label.set_rotation(0)\n",
    "    \n",
    "for idx in [3,9,10]:\n",
    "    plt.hlines(idx, *plt.xlim(), color='black', linewidth=4)\n",
    "    \n",
    "plt.text(1.02, 0.91, \"Most Expressed\", va='center', ha='left', fontsize=24, fontweight='bold', rotation=90, transform=plt.gca().transAxes)\n",
    "plt.text(1.02, 0.58, \"Largest Difference Samples\", va='center', ha='left', fontsize=24, fontweight='bold', rotation=90, transform=plt.gca().transAxes)\n",
    "plt.text(1.02, 0.18, \"Most SNVs\", va='center', ha='left', fontsize=24, fontweight='bold', rotation=90, transform=plt.gca().transAxes)\n",
    "plt.title(\"Gene Expression Ordered by Likelihood of Cell Being in the Metastasis Branch\", fontsize=50, pad=25)\n",
    "plt.ylabel(\"Genes\", fontsize=26, fontweight='bold')\n",
    "plt.xlabel(\"Cells\", fontsize=26, fontweight='bold', labelpad=10)\n",
    "# plt.savefig(r\"../data/results/figures/gene_expression.pdf\", format=\"pdf\")\n",
    "plt.show()"
   ],
   "id": "3932ef8344e3727",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "r = 2\n",
    "path = rf\"../data/results/{study_num}/sciterna\"\n",
    "reference = pd.read_csv(rf'../data/input_data/{study_num}/ref.csv', index_col=0)\n",
    "path_parent = os.path.join(path, \"sciterna_parent_vec\", f\"sciterna_parent_vec_{r}r0.txt\")\n",
    "path_mut_loc = os.path.join(path, \"sciterna_mutation_location\", f\"sciterna_mutation_location_{r}r0.txt\")\n",
    "path_selected = os.path.join(path, \"sciterna_selected_loci\", f\"sciterna_selected_loci_{r}r0.txt\")\n",
    "parent_vec = np.loadtxt(path_parent, dtype=int)\n",
    "mut_locs = np.loadtxt(path_mut_loc, dtype=int)\n",
    "selected_mutations = np.loadtxt(path_selected, dtype=int)\n",
    "\n",
    "selected_loci = reference.columns[selected_mutations]\n",
    "selected_genes = convert_location_to_gene(selected_loci)\n",
    "\n",
    "ctsb_indices = [n for n, c in enumerate(selected_genes) if \"CTSB\" in c]"
   ],
   "id": "c75e30b3009f65aa",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "n_cells = int(((len(parent_vec)+1)/2))\n",
    "ct = CellTree(n_cells=n_cells, n_mut=0)\n",
    "ct.use_parent_vec(parent_vec)\n",
    "ct.mut_loc = mut_locs[ctsb_indices]\n",
    "\n",
    "graph = ct.to_graphviz(gene_names=[\"C\" for _ in range(3000)])\n",
    "graph.attr(dpi='50')\n",
    "graph.attr(rankdir='LR')\n",
    "\n",
    "color_row = filtered_df_top_genes.loc[\"CTSB\"].values\n",
    "color_row = color_row / np.max(color_row)\n",
    "\n",
    "cmap = plt.get_cmap('Greens')\n",
    "colors = cmap(color_row)\n",
    "\n",
    "for n in range(int((len(parent_vec)+1)/2)):\n",
    "    rgba_color = colors[n]\n",
    "    hex_color = mcolors.to_hex(rgba_color)\n",
    "\n",
    "    if n < 65: # primary\n",
    "        prob = (n_bootstrap - posterior_node_after_branching[n]) / 500\n",
    "        graph.node(str(n), label=\"\", shape='circle', style='filled', color=\"red\", \n",
    "                   fillcolor=hex_color, fixedsize=\"true\", width=str(prob), height=str(prob), penwidth=\"10\")\n",
    "    elif n >= 65: # metastasis\n",
    "        prob = posterior_node_after_branching[n] / 500\n",
    "        graph.node(str(n), label=\"\", shape='circle', style='filled', fillcolor=hex_color, \n",
    "                   color=\"blue\", fixedsize=\"true\", width=str(prob), height=str(prob), penwidth=\"15\")\n",
    "graph.attr(ratio=\"0.23\") \n",
    "\n",
    "graph_file = \"../data/results/figures/representative_tree\"\n",
    "graph.render(graph_file, format='pdf', cleanup=True)"
   ],
   "id": "8b4adc20116d9199",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fig, ax = plt.subplots(figsize=(6, 1))\n",
    "norm = mcolors.Normalize(vmin=0, vmax=1)\n",
    "sm = cm.ScalarMappable(cmap='Greens', norm=norm)\n",
    "sm.set_array([])\n",
    "\n",
    "cbar = fig.colorbar(sm, ax=ax, orientation='horizontal', fraction=2, pad=0, aspect=5)\n",
    "cbar.set_label('Gene Expression CTSB', fontsize=24)\n",
    "cbar.ax.tick_params(labelsize=10)\n",
    "ax.axis('off')\n",
    "colorbar_path = os.path.join(path, \"gene_expression_colorbar.svg\")\n",
    "plt.show()"
   ],
   "id": "8b51a2c5404f711c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "c09d04b7fc44fd62",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
